src/CMakeLists.txt

```
@@ -0,0 +1,77 @@
+# set minimum cmake version
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+
+# Uncomment for VERBOSE Makefiles. Useful for debugging.
+set(CMAKE_VERBOSE_MAKEFILE ON)
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+# project name and language
+project(ImpactT LANGUAGES Fortran)
+
+# BUILD_SHARED_LIBS is a global flag offered by CMake
+# to toggle the behavior of add_library
+# With OFF it will create a static library
+set(BUILD_SHARED_LIBS OFF)
+
+option(USE_MPI "Activate MPI build" OFF)
+
+
+list(APPEND _sources
+    DataStruct/NumConst.f90
+    DataStruct/PhysConst.f90
+    DataStruct/Pgrid.f90
+	DataStruct/Data.f90
+	Func/Timer.f90
+	Func/Transpose.f90
+	Func/Fldmger.f90
+	Func/Ptclmger.f90
+	Func/FFT.f90
+    Appl/BPM.f90
+    Appl/CCL.f90
+    Appl/CCDTL.f90
+    Appl/DTL.f90
+    Appl/SC.f90
+    Appl/Multipole.f90
+	Appl/DriftTube.f90
+	Appl/Quadrupole.f90
+	Appl/ConstFoc.f90
+	Appl/SolRF.f90
+	Appl/Sol.f90
+	Appl/Dipole.f90
+	Appl/EMfld.f90
+	Appl/EMfldCart.f90
+	Appl/EMfldCyl.f90
+	Appl/EMfldAna.f90
+	Appl/BeamLineElem.f90
+	Appl/CompDom.f90
+	Appl/BeamBunch.f90
+	Appl/Depositor.f90
+	Appl/Ranger.f90
+	Appl/Field.f90
+	Appl/Distribution.f90
+	Contrl/Input.f90
+	Contrl/Output.f90
+	Contrl/AccSimulator.f90
+	Contrl/main.f90
+)
+
+set(OUTPUT_NAME "ImpactTexe")
+
+if(USE_MPI)
+    find_package(MPI REQUIRED)
+    add_definitions(${MPI_Fortran_COMPILE_FLAGS} -DUSE_MPI)
```

Avoid global functions like `add_definitions` and `include_directories` and use targets instead:
Move the `add_executable` further up in the file and then you can do:
```suggestion
    target_compile_definitions(${OUTPUT_NAME} PUBLIC ${MPI_Fortran_COMPILE_FLAGS})
    target_compile_definitions(${OUTPUT_NAME} PUBLIC USE_MPI)
```

Specifically for MPI, I would do this though:
```suggestion
    target_compile_definitions(${OUTPUT_NAME} PUBLIC USE_MPI)
    target_link_libraries(${OUTPUT_NAME} PUBLIC MPI::MPI_Fortran)
```

---

In case you need to add sources depending on options, move `add_executable` up and add sources later via:
```cmake
add_executable(${OUTPUT_NAME})
# ...
target_sources(${OUTPUT_NAME} PRIVATE ${_sources})
```

---

That was great! I switched to use this way! Since we are changing the `OUTPUT_NAME` depending on MPI or not I have a duplicated `    add_executable(${OUTPUT_NAME})`. Any ideas how to solve that?

---

Awesome! I added a little suggestion to simplify this :)